### ISSUE TITLE
adding a delay after using merchant

### ISSUE THREAD

**Issue by krazy03** [kolton/d2bot-with-kolbot#581]
So the sticky issue with NPC's (from what i've seen) Since I can sort of replicate it when playing manually is that when you close NPC window (this can happen with trade window and stash too) and then try to click (move) somewhere and both actions were done almost at the same time its when you get stuck in the endless loop (game thinks stash /  trade window / npc window is up) and you have to wait until your bot times out or fails each script and exits game. 

TL:DR

Where can I add (which script) a 3 or 5 second delay after bot closes NPC / Stash window to start new action? Basically sell > close window > wait 3-5 seconds then continue?

Thanks.

**Comment by DarkHorseDre** [comment]
I'm not sure I've seen the problem, but I'd check town.js and shopbot scripts as a start

**Comment by spark10z** [comment]
i got same prob..
https://github.com/kolton/d2bot-with-kolbot/issues/680

do we need such big delay like 3000-5000 msec ?

**Comment by DarkHorseDre** [comment]
I just had a quick look at town.js and #680 - seems ping could play a big part.  
e.g. `delay(1000 + me.ping`, and general lag issues..

Do any of you experiencing this have sub 100 ping?  what is your average ping level?

**Comment by spark10z** [comment]
yea.. PING must be the reason .. because of my location, its big (200-400) and i cant help itself.   so delay() maybe.
Ill try delay(1000 + me.ping);  later.   thanks.

**Comment by DarkHorseDre** [comment]
No bro, I was showing you an example entry from town.js - it means that the script is written to be dynamic according to ping.. I cant test as my bots have been down a while and will be until I get some time.
You need to work out if the delays() and/or pings need tweaking to avoid your issue.

btw - do you ever get stuck when playing manually?   give it a try and see. if not, then the scripts probably need tweaking to suit your lag.

**Comment by spark10z** [comment]
hi, thank you for quick response. I have same lag lock like problem even on manual play. but far less and bit different from bots. manual problem is mostly happen at stash while dropping item from stash, the item sticks to pointer and can not drop or click neither, need rejoin game in this case. also more rare, but shop lag lock happens too when I try to leave NPC too fast.

in town.js,  I have already put many <delay( 1000 + me.ping * 2);> to the place where just after each task. also there were same sort of delays like < 750 + me.ping> , so changed number bigger.  I did not do in each function side, but I put them in calling function side.  and still getting problem.

I will move delay scripts line to each function side for next test, while checking each interaction with NPC more specifically.  any suggestion ?

**Comment by DarkHorseDre** [comment]
No worries.

I guess you need to find the place where the issue occurs.  I don't know much about the cause of these things but I would use the `print()` or `say()` commands to put lines in the chat/console before and after each task so that you see when the problem occurs (the lag will stop the text which should display next from displaying, so you'll know the action before that is where it's stuck).  Actually, it's better to not print in game chat as your account can get muted, so write to the console (which you see when you press the `home` key).

For example: I wrote the following line when troubleshooting failed to use wp error in mfhelper (I forget which prints to the console, `say()` or `print()` ):
`print("Failed to use portal.[275 - in town]");` - 275 is the line number of the code block so I could find it (as there were many similar blocks taking wp to town)

**Comment by spark10z** [comment]
yea.. i need to do that.. thank you for clear explanation.

**Comment by krazy03** [comment]
I wasn't able to get it fixed properly, but its not a ping issue for me. I have constant 72-96 Ping, sometimes spikes to 130ish but i think thats totally normal. In my opinion anyone can get it, just close menu + move command somewhere at almost same time fast and it'll happen. Can be at stash, trade window or NPC, anything that opens a window up.

**Comment by spark10z** [comment]
hi krazy, thank you for info.
even you say its not ping issue, it seems there is a big difference between low ping and high ping to the possibility to get the lock. as i wrote my first post here, https://github.com/kolton/d2bot-with-kolbot/issues/680
on ping 200, i get one lock in 7-10 games(30min each), and on ping 400, i get it in 1.5 games.. lol.  i can not continue mf runs.. in act1, at first doChore, char get caught often.
i ll inspect more.

**Comment by DarkHorseDre** [comment]
ok so it sounds like the scripts are closing the window and moving/doing next click too fast - I've seen this a few times but not much (my ping is always under 100, from memory).

I don't have time but i think between you all you can comment the scripts like I mentioned in previous post/or find the actions in the scripts and work out when the clicks occur, and place longer delays between them and test.

**Comment by spark10z** [comment]
yes, i will put delay at after each points of opened window closing.. im bit busy right now, maybe in few days.  thanks again dark.

**Comment by spark10z** [comment]
hi again, it seems noah and guys fixed this prob.. i m trying it now.
check my original thread please.
https://github.com/kolton/d2bot-with-kolbot/issues/680

**Comment by justinpees** [comment]
maybe you need to edit the settings within kolbot and set the delay timer to the max offered there? not sure if thats what you need. very vague. if you literally mean everytime you visit a merchant then that would be done in town.js file. just add a delay there when merchants are being interacted.
